### PR TITLE
Add forecasting predictor

### PR DESCRIPTION
**Pull request by yx1215**
*Issue #, if available:*

*Description of changes:*
Add forecasting predictor.
#
This PR is used for further discussion about add the forecasting task to Autogluon, which should be parallel to tasks such as tabular prediction and image classifications etc, replacing  #684 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Output in ```simple_forecasting_example.py```:
```
    model     score  fit_order
0     SFF -0.617567          2
1   MQCNN -0.647109          1
2  DeepAR -0.776917          3

0.3063789898931557
```

Output in ```advanced_forecasting_example.py```:
```
              model     score  fit_order  test_score
0       SFF/trial_4 -0.771542          5   -0.455381
1       SFF/trial_5 -0.800340          6   -0.469165
2     MQCNN/trial_0 -0.859082          1   -0.819518
3     MQCNN/trial_1 -0.859113          2   -0.819403
4    DeepAR/trial_2 -0.882853          3   -0.894841
5    DeepAR/trial_3 -0.899381          4   -0.917463
6  SFF/trial_4_FULL       NaN          7   -0.559157

0.4553322822285291

                    0.5
2020-04-22   509.987183
2020-04-23   617.542175
2020-04-24  1040.601807
2020-04-25  1154.234009
2020-04-26  1073.659058
2020-04-27  1090.976196
2020-04-28   696.362183
2020-04-29  1438.374390
2020-04-30  1149.915405
2020-05-01   468.797913
2020-05-02   889.827393
2020-05-03   938.827820
2020-05-04   557.264038
2020-05-05  1205.893555
2020-05-06   379.958221
2020-05-07   616.763733
2020-05-08   877.277222
2020-05-09  1134.502930
2020-05-10   620.372009
```
Things to improve in the future:
- Ensembling 
- Dynamic features
- More detailed preprocessing for dynamic/static features.(currently the code only do auto inferring of real/cat features and fillna, other things like standardizing might also be useful)
- Making plots for predictions
- Subsampling dataset(select a range from given train data)
- Improve ways to do validation
- New cat-static features in prediction might lead to failure.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1198-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/1/index.html

**Comment by szha** [comment]
Job PR-1198-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/2/index.html

**Comment by szha** [comment]
Job PR-1198-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/3/index.html

**Comment by szha** [comment]
Job PR-1198-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/4/index.html

File: forecasting/tests/unittests/test_forecasting.py [hunk 1]
```diff
@@ -1,4 +1,54 @@
-import autogluon.forecasting
+import pandas as pd
+from autogluon.forecasting import ForecastingPredictor
+from autogluon.forecasting import TabularDataset
+import autogluon.core as ag
+
 
 def test_forecasting():
-	print("Not Implemented")
+    train_data = TabularDataset("https://autogluon.s3-us-west-2.amazonaws.com/datasets/CovidTimeSeries/train.csv")
```
Comment (jwmueller) [inline forecasting/tests/unittests/test_forecasting.py:8]: Add the other tests we discussed

Comment (jwmueller) [reply forecasting/tests/unittests/test_forecasting.py:8]: Ideally the first test should not have any optional args specified (ie. what we expect most users to run), except what we need to ensure it runs quickly.  Is there any way for the test to run quickly without specifying: 
```
hyperparameters={
                                                                                 "MQCNN": {
                                                                                     'context_length': ag.Int(70, 90,
                                                                                                              default=prediction_length * 4),
                                                                                     "num_batches_per_epoch": 10,
                                                                                     "epochs": 5}
```
                                                           
Perhaps we can subsample the data to be even tinier and then the test will run quickly enough without any optional args like this?

File: forecasting/src/autogluon/forecasting/predictor/predictor.py [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+
+        refit_full: bool, default=False
+            Whether to retrain all models on all of the data (training + validation) after the normal training procedure.
+            If `refit_full=True`, it will be treated as `refit_full='all'`.
+            If `refit_full=False`, refitting will not occur.
+            Valid str values:
+                `all`: refits all models.
+                `best`: refits only the best model (and its ancestors if it is a stacker model).
+
+        set_best_to_refit_full: bool, default=False
```
Comment (jwmueller) [inline forecasting/src/autogluon/forecasting/predictor/predictor.py:231]: Why not make this = True by default?  Don't we want most novice ML users to automatically benefit from this?

**Comment by szha** [comment]
Job PR-1198-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/5/index.html

**Comment by szha** [comment]
Job PR-1198-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/6/index.html

**Comment by szha** [comment]
Job PR-1198-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/7/index.html

**Comment by szha** [comment]
Job PR-1198-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/8/index.html

File: forecasting/src/autogluon/forecasting/predictor/predictor.py (hunk 2 above)
Comment (yx1215) [reply forecasting/src/autogluon/forecasting/predictor/predictor.py:239]: This is now changed to:
it is by default False, when refit is set, it is going to be turned to True by default, but the user can still set their own value to override it.

**Comment by szha** [comment]
Job PR-1198-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/13/index.html

**Comment by szha** [comment]
Job PR-1198-14 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/14/index.html

**Comment by szha** [comment]
Job PR-1198-16 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/16/index.html

**Comment by szha** [comment]
Job PR-1198-17 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/17/index.html

**Comment by szha** [comment]
Job PR-1198-18 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/18/index.html

**Comment by szha** [comment]
Job PR-1198-19 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/19/index.html

**Comment by szha** [comment]
Job PR-1198-20 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/20/index.html

**Comment by szha** [comment]
Job PR-1198-21 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/21/index.html

**Comment by szha** [comment]
Job PR-1198-22 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/22/index.html

**Comment by szha** [comment]
Job PR-1198-23 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/23/index.html

**Comment by szha** [comment]
Job PR-1198-24 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/24/index.html

**Comment by szha** [comment]
Job PR-1198-25 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/25/index.html

**Comment by szha** [comment]
Job PR-1198-29 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/29/index.html

**Comment by szha** [comment]
Job PR-1198-30 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/30/index.html

**Comment by szha** [comment]
Job PR-1198-31 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/31/index.html

**Comment by szha** [comment]
Job PR-1198-33 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/33/index.html

**Comment by szha** [comment]
Job PR-1198-34 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/34/index.html

**Comment by szha** [comment]
Job PR-1198-35 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/35/index.html

**Comment by szha** [comment]
Job PR-1198-36 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/36/index.html

**Comment by szha** [comment]
Job PR-1198-37 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/37/index.html

**Comment by szha** [comment]
Job PR-1198-38 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/38/index.html

**Comment by szha** [comment]
Job PR-1198-42 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/42/index.html

**Comment by szha** [comment]
Job PR-1198-43 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/43/index.html

**Comment by szha** [comment]
Job PR-1198-44 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/44/index.html

**Comment by szha** [comment]
Job PR-1198-45 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/45/index.html

**Comment by szha** [comment]
Job PR-1198-46 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/46/index.html

**Comment by motin** [comment]
@yx1215 What a great PR this is! I have tried the code and apart from not being able to leave `index_column` to None (as hinted in the docs should be possible), I haven't run into any bugs that I know of. 

I do miss the ability to use the plotting helpers in GluonTS (https://ts.gluon.ai/tutorials/forecasting/quick_start_tutorial.html#Visualize-and-evaluate-forecasts) out of the box - some convenience methods for exposing those after training would be helpful.

Do you think you will include dynamic features as well as static features, or is that left for future improvements once this gets merged?

Again, great work! I hope this gets merged in the near future :)

**Comment by yx1215** [comment]
> @yx1215 What a great PR this is! I have tried the code and apart from not being able to leave `index_column` to None (as hinted in the docs should be possible), I haven't run into any bugs that I know of.
>  I can check later what happens if index_column=None, but for now you might always want to set index_column to a specific value. 
> I do miss the ability to use the plotting helpers in GluonTS (https://ts.gluon.ai/tutorials/forecasting/quick_start_tutorial.html#Visualize-and-evaluate-forecasts) out of the box - some convenience methods (or docs) for visualization accuracy after training would be helpful.
> Yes visualization methods are on the calendar as said in the descriptions, but still we would want to first get the whole PR merged before we develop more features.
> Do you think you will include dynamic features as well as static features, or is that left for future improvements once this gets merged?
> For the dynamic feature things, since the use of dynamic features is a little bit complicated in gluonts. For example, some models can predict without dynamic features even if it is given for training, but some does not. So we would want to add the dynamic features later as the current PR is already a very large one. 
> Again, great work! I hope this gets merged in the near future :)

**Comment by szha** [comment]
Job PR-1198-49 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/49/index.html

**Comment by szha** [comment]
Job PR-1198-50 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/50/index.html

**Comment by szha** [comment]
Job PR-1198-51 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/51/index.html

**Comment by szha** [comment]
Job PR-1198-52 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/52/index.html

**Comment by szha** [comment]
Job PR-1198-55 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/55/index.html

**Comment by szha** [comment]
Job PR-1198-56 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/56/index.html

**Comment by szha** [comment]
Job PR-1198-63 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1198/63/index.html

**Comment by Innixma** [comment]
This PR is continued in #1538.

**Pull request closed without merging**